### PR TITLE
Fix product rating aggregates lagging on REST review changes

### LINKED ISSUE #29906: REST API: Average Product Rating Doesn't Show Up if There Is Only One Review
This was reported by a user on #3965783-zen.

**Prerequisites (mark completed items with an \[x\]):**

- [ ] I have have carried out troubleshooting steps and I believe I have found a bug.
- [X] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

**Describe the bug**

The user was using the API to submit reviews via an app they're developing.

This is the API URL that the user called that showed this problem (I’ve changed the sensitive information):

```
https://usersite.com/wp-json/wc/v3/products/reviews?review=this is review body4&title=this is title11&rating=5&product_id=1647&reviewer=username&reviewer_email=user@usersite.com
```

**Expected behavior**<br>The average rating should be updated.

**Actual behavior**<br>The review was created but the average wasn’t updated.

**Screenshots**

Markup on 2021-05-14 at 19:43:13

**Isolating the problem (mark completed items with an \[x\]):**

- [X] The user has deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](<https://woocommerce.com/storefront/>).
- [ ] I can reproduce this bug consistently using the steps above.

**User's Findings and Suggested Fix**

The user's solution was to call this action in their API call after creating the review:

```
do_action( 'wp_update_comment_count', $p['product_id']);
```

It calls the clear_transients method which resets all averages and arrays.

The user further investigated why that action wasn't being called by default. They found that, when a comment is created on the website, and not through the API, the following action is supposed to be triggered:

```
add_action( 'comment_post', array( __CLASS__, 'add_comment_rating' ), 1 );
```

However, this action is not called by the API.

### PR DESCRIPTION
**Pull request by oaratovskyi**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

REST-created and updated product review aggregates lag one operation because WordPress and the REST controllers previously recomputed counts before the new rating was available. With N REST-created reviews, the aggregate therefore reflected N-1 ratings ([issue context](https://github.com/woocommerce/woocommerce/issues/29906#issuecomment-5285559679)). Moving a review through wc/v3 also refreshed the destination product but left the source stale.

This PR:

- Refreshes aggregates after approved rated creates.
- Passes rating meta through comment updates so the existing count callback sees the new value.
- Recounts each affected product exactly once after request- or filter-driven review moves.
- Rejects malformed preprocess-filter output and invalid wc/v3 destination IDs with REST errors.
- Removes API-test workarounds that manually recomputed product aggregates.

<details>
<summary>Implementation details</summary>

- Create requests refresh aggregates directly with `WC_Comments::clear_transients()` after rating meta is stored. Reviews without a rating and held reviews skip this work because they cannot affect approved rating aggregates.
- Update requests include integer rating meta in `wp_update_comment()`, whose existing count callback then saves the current aggregates once. This covers wc/v1, wc/v2, wc/v3, `/batch`, `wc-analytics/products/reviews`, and `wp wc product_review`.
- Review moves capture the product Core will count before `wp_update_comment_data` runs, then re-read the persisted review and call `wp_update_comment_count()` for each affected product Core missed. This keeps core counts, WooCommerce aggregates, hooks, and deferred comment counting aligned without duplicate recounts.
- Both public preprocess-filter results are validated before array access. Nested `comment_meta` must also be an array, and wc/v3 validates that an explicitly supplied destination is a positive product ID.

</details>

**Backward compatibility:** No PHP signature, REST route, schema, or hook-argument changes. Observable behavior changes are intentional:

- A rating-only wc/v1 or wc/v2 `PUT` now returns 200 and stores the rating instead of returning 500.
- A genuine v1/v2 comment update failure now returns 500 without writing rating meta instead of returning 200 after writing it.
- Invalid wc/v3 destinations return `woocommerce_rest_product_invalid_id` instead of detaching the review. This includes `comment_post_ID = 0` from either request input or the `woocommerce_rest_preprocess_product_review` filter, so filter consumers that intentionally detached reviews through zero now receive a 404.
- Malformed public preprocess-filter output returns the controller's existing REST update error instead of causing a PHP error.
- Rated updates save the product once. Approved rated creates retain the required second product save; held creates skip it.
- Moving a review refreshes each affected product through `wp_update_comment_count()`, so callbacks on its post-update hooks fire and deferred counting is honored. Each product is counted exactly once even when `wp_update_comment_data` redirects the persisted destination. `product.updated` webhooks remain deduplicated.

Existing stale aggregates are not backfilled; they repair on the next recompute. This adds no site-topology or path assumptions. Multisite was not separately tested.

**Out of scope:** `rating: 0` remains a successful no-op on update. Batching the repeat create-side save on `/batch` is also unchanged.

Closes #29906.

Bug introduced in PR #27100.

### Screenshots or screen recordings:

No production UI changes.

### How to test the changes in this Pull Request:

Run these commands in a WP-CLI shell connected to a store running this branch.

1. Create a source product and two rated reviews through the REST-backed WP-CLI commands:

    ```sh
    SOURCE_PRODUCT_ID="$(wp wc product create --name='Rating source' --regular_price=10 --user=1 --porcelain)"
    FIRST_REVIEW_ID="$(wp wc product_review create "$SOURCE_PRODUCT_ID" --review='Solid.' --name='Tester One' --email='tester-one@example.com' --rating=5 --user=1 --porcelain)"
    SECOND_REVIEW_ID="$(wp wc product_review create "$SOURCE_PRODUCT_ID" --review='Needs work.' --name='Tester Two' --email='tester-two@example.com' --rating=1 --user=1 --porcelain)"
    wp post meta get "$SOURCE_PRODUCT_ID" _wc_average_rating
    wp post meta get "$SOURCE_PRODUCT_ID" _wc_rating_count --format=json
    ```

    On `trunk`, the average lags at `5.00`. This branch returns `3.00` with counts `{"1":1,"5":1}`.

2. Update only the second rating through wc/v1:

    ```sh
    wp eval "\$request = new WP_REST_Request( 'PUT', '/wc/v1/products/${SOURCE_PRODUCT_ID}/reviews/${SECOND_REVIEW_ID}' ); \$request->set_body_params( array( 'rating' => 3 ) ); \$response = rest_do_request( \$request ); echo \$response->get_status();" --user=1
    wp comment meta get "$SECOND_REVIEW_ID" rating
    wp post meta get "$SOURCE_PRODUCT_ID" _wc_average_rating
    ```

    On `trunk`, the request returns 500, leaves the rating at `1`, and recalculates the stale average to `3.00`. This branch returns 200, stores rating `3`, and updates the average to `4.00`.

3. Create a destination and move the second review through wc/v3:

    ```sh
    DESTINATION_PRODUCT_ID="$(wp wc product create --name='Rating destination' --regular_price=10 --user=1 --porcelain)"
    wp eval "\$request = new WP_REST_Request( 'PUT', '/wc/v3/products/reviews/${SECOND_REVIEW_ID}' ); \$request->set_body_params( array( 'product_id' => (int) ${DESTINATION_PRODUCT_ID} ) ); \$response = rest_do_request( \$request ); echo \$response->get_status();" --user=1
    wp post get "$SOURCE_PRODUCT_ID" --field=comment_count
    wp post meta get "$SOURCE_PRODUCT_ID" _wc_review_count
    wp post meta get "$SOURCE_PRODUCT_ID" _wc_average_rating
    wp post get "$DESTINATION_PRODUCT_ID" --field=comment_count
    wp post meta get "$DESTINATION_PRODUCT_ID" _wc_review_count
    wp post meta get "$DESTINATION_PRODUCT_ID" _wc_average_rating
    ```

    The request returns 200. The source core and WooCommerce review counts are both `1`, with an average of `5.00`. The destination core and WooCommerce review counts are both `1`, with an average of `3.00`.

4. Remove the temporary data:

    ```sh
    wp comment delete "$FIRST_REVIEW_ID" "$SECOND_REVIEW_ID" --force
    wp post delete "$SOURCE_PRODUCT_ID" "$DESTINATION_PRODUCT_ID" --force
    ```

### Testing that has already taken place:

- **Automated:** The wc/v1, wc/v2, and wc/v3 product-review controller suites pass with 40 tests and 173 assertions. The focused `wp_update_comment_data` redirect regression passes in wc/v1 and wc/v3 with 2 tests and 19 assertions, proving the source and persisted destination are each recounted exactly once. PHPStan reports no errors for both modified controllers. Both required PHP lint checks pass; full branch lint exits successfully with three unchanged JavaScript warnings.
- **Manual:** WordPress 7.0.4 with PHP 8.1. Verified rated creates, a rating-only wc/v1 update, a wc/v3 move, storefront aggregates, and product-review admin editing. All probe data was removed afterward.

### Milestone

- [x] Automatically assign milestone for the **[next WooCommerce version](../blob/trunk/plugins/woocommerce/woocommerce.php#L6)**

> **Note:** Check the box above to have the milestone automatically assigned when merged.
> Alternatively (e.g. for point releases), manually assign the appropriate milestone.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [x] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment

Created manually.

</details>

### Release Communication

Select the release communication labels this PR needs:

- [ ] **Feature Highlight** - For user-facing features (what changed, user impact)
- [ ] **Developer Advisory** - For developer-facing changes (what changed, how to detect, actions needed)

<details>
<summary>When to use each?</summary>

**Feature Highlight**: New features, UI changes, or improvements that merchants/store owners will notice.

- Example: "New bulk editing for products", "Improved checkout performance"

**Developer Advisory**: Breaking changes, deprecations, or changes that affect themes/plugins/extensions.

- Example: "Hook signature change", "Deprecated filter", "REST API field removed"

Selected labels are applied when the pull request is merged.
</details>

### Use of AI Tools

Claude Code helped review the original fix and draft the initial implementation and description. Codex implemented follow-ups, ran the final local verification, and revised the PR body. The author will review the final implementation, test results, and PR body before merge.

### CODE REVIEW

File: plugins/woocommerce/tests/php/includes/rest-api/Controllers/Version1/class-wc-rest-product-reviews-v1-controller-tests.php [hunk 1]
```diff
@@ -170,4 +170,70 @@ public function test_cannot_delete_other_comment_types() {
...
+		$product_id = ProductHelper::create_simple_product()->get_id();
+
+		$response = $this->create_review( $product_id, 'Holds up to daily use.', 5 );
+		$this->assertEquals( 201, $response->get_status(), 'The review is created successfully.' );
+
+		$product = wc_get_product( $product_id );
+		$this->assertEquals( 5, $product->get_average_rating(), 'The average rating includes the review created in the same request.' );
+		$this->assertEquals( array( 5 => 1 ), $product->get_rating_counts(), 'The rating counts include the review created in the same request.' );
+		$this->assertEquals( 1, $product->get_review_count(), 'The review count includes the review created in the same request.' );
+	}
```
[the remark this answers is not in the document]
Comment (oaratovskyi) [reply plugins/woocommerce/tests/php/includes/rest-api/Controllers/Version1/class-wc-rest-product-reviews-v1-controller-tests.php:188]: You are right, added it in `439f9543c0b`.

The test hooks `rest_pre_insert_product_review`, points `comment_post_ID` at a second product, then checks that product's average rating, rating counts and review count.

I checked that it actually catches the bug. With v1 reverted to recomputing the product named in the request, it fails with `Failed asserting that '0' matches expected 5`. With the fix it passes.

One thing I left out on purpose. I did not assert that the requested product stays untouched. Before the fix that product gets recomputed to 0, and after the fix it is never touched, so it reads 0 either way and the assertion would prove nothing.

**Comment by mordeth** [comment]
Approach is right, three accuracy problems.

**1 (rating-only v1/v2 edits return 500, never reaching the new recompute):** on a `PUT` changing only `rating`, no comment-row field differs, so `wp_update_comment()` returns `0` and v1:362 turns that into a 500 above the new `clear_transients()` block. The rating is never stored. Confirmed by running it. v3 is correct at :596 with `false === $updated`. The new v1 test passes only because it also sets `review`. Pre-existing guard, so not a regression, but the v1/v2 half of the fix never fires on its simplest case.

**2 (changelog overstates):** it claims rating edits and moves are fixed via REST. Neither holds for v1/v2. Scope it to v3, or make v1/v2 deliver it.

**3 (the comment at v1:374 is false):** `get_review()` guarantees ownership only until `prepare_item_for_database()` applies `rest_preprocess_product_review` at :529, which can rewrite `comment_post_ID`. Reproduced with a redirecting callback: destination reads 3.00 against a true 1.50. Create defends against this at :320 with a test, v3 update re-reads the comment, so v1/v2 update is the one path left out.

**Proposal:** relax `0 === $updated` to `false === $updated` on v1, and pass the rating through the update:

```php
$prepared_review['comment_meta']['rating'] = $request['rating'];
```

`wp_update_comment()` writes `comment_meta` before `wp_update_comment_count()`, so core's own recompute sees it. Fixes 1, drops the extra save on update, covers most of 3. Both must land together: `comment_meta` alone persists the rating and recomputes behind the 500. Then a rating-only test, the comment, the changelog.

**Follow-ups (not blocking):**

- Rated writes save the product twice (1 on trunk, 2 here). Disclosed; `comment_meta` fixes the update half.
- A rated create with `status: "hold"` saves where trunk did not. Aggregates filter `comment_approved = '1'`, so gate on it.
- Direct `clear_transients()` bypasses `wp_defer_comment_counting()`. `wp_update_comment_count( $original_product_id )` at v3:628 honors it, stays removable, and fixes `wp_posts.comment_count`, stale on the source after a move.
- Nothing pins the `rating: 0` decision.

**Comment by oaratovskyi** [comment]
Thanks for the careful review. You are right on all three points. I ran them over REST against a local site rather than relying only on PHPUnit.

**1.** Confirmed. Before these follow-ups, `PUT /wc/v1/products/<id>/reviews/<id>` with `{"rating": 2}` returned 500 and left the rating untouched. Adding `review` to the same body returned 200, which is exactly why the earlier v1 test passed. Trunk behaves the same, so it predates this PR, but you are right that it left the v1/v2 half of the fix unreachable in its simplest case. This branch now returns 200 for the rating-only request. In the final local check it stored rating 3 and moved the average from 3.00 to 4.00.

**2.** You are right that the changelog overstated route-level move support. A REST move is a wc/v3 feature. The shared v1/v2 controller only repairs a source move if `rest_preprocess_product_review` changes the comment. I narrowed the entry so it says creation and re-rating apply across REST, while route-level moves are wc/v3.

**3.** Confirmed, and your numbers landed on the nose. With a redirecting callback, the destination read 3.00 against a true 1.50, with counts `{"1":1,"5":1}`. The comment was wrong. The v1/v2 path now captures the source, passes the integer rating through `comment_meta`, re-reads the stored comment, and refreshes the source only when the review actually moved.

**The proposal.** It works. Passing the rating through update-side `comment_meta` lets WordPress's existing count callback see the new value. The v1 guard now treats only `false` as failure, and the existing `! empty( $request['rating'] )` guard prevents an omitted or zero rating from wiping meta. Dropping the explicit destination refresh also removes the extra product save. Your follow-up about that double save caught the same issue in v3:

| v3 rated update | Product saves |
| --- | --- |
| trunk | 1 |
| PR at review time | 2 |
| after these follow-ups | 1 |

**Follow-ups.** I took the hold gate. On a default site with a first-time reviewer, trunk saves the product zero times and the reviewed branch saved it once, while the held rating is excluded either way. The final v1 and v3 tests now observe zero unnecessary saves for held creates.

I also pulled `wp_update_comment_count()` into this PR for the source product. The reproduction left `_wc_review_count` at 1 and `wp_posts.comment_count` at 2. The final local wc/v3 move returned 200 and left both source counts at 1, with the destination core count at 1. The new tests also cover source hook order and deferred comment counting.

`rating: 0` now has a successful no-op test in v1, v2, and v3.

File: plugins/woocommerce/includes/rest-api/Controllers/Version1/class-wc-rest-product-reviews-v1-controller.php [hunk 2]
```diff
@@ -343,18 +356,41 @@ public function update_item( $request ) {
...
+
+		/*
+		 * Zero means no comment-table row changed; false is the actual update failure. A rating-only
+		 * request legitimately returns zero while still storing comment_meta and refreshing counts.
+		 */
 		$updated = wp_update_comment( $prepared_review );
-		if ( 0 === $updated ) {
+		if ( false === $updated ) {
 			return new WP_Error( 'rest_product_review_failed_edit', __( 'Updating product review failed.', 'woocommerce' ), array( 'status' => 500 ) );
 		}
```
[the remark this answers is not in the document]
Comment (oaratovskyi) [reply plugins/woocommerce/includes/rest-api/Controllers/Version1/class-wc-rest-product-reviews-v1-controller.php:390]: Thanks for flagging this. I confirmed the v1 half with a disposable REST
probe: `literal\path \\server` was stored as `literalpath \server`, so the
long-standing v1 convention can lose literal backslashes.

I am not changing it in this PR because the new value introduced here is the
integer rating, which is unaffected, and matching v3 is not a complete
behavioral oracle. v3 sanitizes review text with `wp_filter_post_kses` before
the controller and also did not preserve the probe input byte-for-byte. The
requested v3 comment change is incorrect as well: `wp_slash()` delegates to
`addslashes_strings_only()`, so an integer remains an integer. I am keeping the
accurate comment and recording the v1 string behavior as a separate regression
that needs its own route-level test.

File: plugins/woocommerce/tests/php/includes/rest-api/Controllers/Version3/class-wc-rest-product-reviews-controller-tests.php [hunk 3]
```diff
@@ -200,4 +200,356 @@ public function test_cannot_delete_other_comment_types() {
...
+		try {
+			$response = $this->server->dispatch( $request );
+		} finally {
+			remove_action( 'woocommerce_update_product', $count_save );
+		}
+
+		$this->assertSame( 201, $response->get_status() );
+		$this->assertSame( 0, $saves );
+		$this->assertEquals( 0, wc_get_product( $product_id )->get_average_rating() );
+	}
```
[the remark this answers is not in the document]
Comment (oaratovskyi) [reply plugins/woocommerce/tests/php/includes/rest-api/Controllers/Version3/class-wc-rest-product-reviews-controller-tests.php:270]: Thanks. I reproduced the edge case, but not the proposed invariant. With a
first-time reviewer, the current test is accurate: held creation causes zero
product saves. With an email that already has an approved comment, core first
approves the insertion and the requested status then moves it to hold. Those
two core count transitions produced two product saves before the controller's
direct-refresh guard.

The held review still stores its rating meta; aggregates exclude it. Avoiding
the core saves would require moving the requested status into insertion and
changing long-standing hook and count ordering, so that needs separate
backward-compatibility and design review. This PR's optimization is deliberately
narrower: skip only the new direct post-meta save for held creates. I am not
changing that behavior here.

**Comment by oaratovskyi** [comment]
I also checked the collapsed v2 nitpick in this review. It is valid: after `rating: 0`, the test could reassert that the average remains 3, matching the v1 and v3 coverage. This is a test-coverage improvement, not a production defect. v2 inherits the same implementation and its route test already pins the 200 response and unchanged rating meta. I recorded the aggregate assertion as a non-blocking follow-up rather than restarting CI solely for the additional assertion.

**Comment by oaratovskyi** [comment]
Thanks for the new-head review, @coderabbitai. I checked this against the controller history and am pushing back on this finding for the current PR.

The v1 create and update filters, and the v3 create filter, have passed their final prepared destination directly to WordPress persistence since 2019. Adding post-filter product validation would newly reject extension-selected outcomes that have worked for about seven years, so it needs a separate, backward-compatibility-documented hardening decision across all legacy review filters.

The aggregate work in this PR is safe with the existing contract. Create-side refreshes target the persisted destination, and `WC_Comments::clear_transients()` returns immediately for post zero and non-products. The v1 update path re-reads the persisted comment and repairs the original source whenever the destination changed, including a move to post zero. Valid cross-product filter rerouting also remains covered by regression tests.

I am keeping the zero-destination source-repair test. If WooCommerce wants to guarantee that filtered reviews always remain attached to products, I agree it should be handled consistently in a dedicated compatibility review rather than changing this one filter path in this PR.

**Comment by mordeth** [comment]
Verified red→green on a real install: a rating 5 create gives `0` on base and `5.00` here, and the rating-only `wc/v2` PUT gives 500 then 200. Three things left.

**1 (`product_id: 0` still orphans the review when the global `$post` is a product):** `get_post_type( 0 )` falls back to `$GLOBALS['post']`, so during an internal `rest_do_request()` from a product context it returns `'product'`, validation passes, and `wp_update_comment()` skips its own check because `! empty( 0 )` is false. Reproduced on head: without the global, 404; with it, 200 and `comment_post_ID=0`. Trunk orphans unconditionally so this is still better, but the BC line saying `product_id: 0` "now returns 404 instead of orphaning the review" is not true in that case. A positive-ID check before `get_post_type()` fixes it; the test needs `$GLOBALS['post']` set to a product.

**2 (non-array `comment_meta` goes from tolerated to 500):** on trunk core ignored it via its own `is_array()` guard and the rating was written separately, so the request succeeded. The `TypeError` the BC note cites only became reachable when this PR added the nested assignment, so it is not the baseline. Normalizing to an empty array before adding the rating keeps trunk's behavior and the new ordering. Same in v1/v2 and v3. Related: v1:371 checks `is_wp_error` but not "is an array", so a filter returning `null` makes `array_key_exists( 'comment_meta', null )` a PHP 8 fatal, and `! is_array( ... )` covers both.

**3 (create-path amplification):** measured, `woocommerce_update_product` per request, base then head: approved rated create 1 then 2, held 0 then 0, unrated 1 then 1. `/batch` at 100 items adds roughly 100 saves, 300 aggregate queries, and 100 extra hook firings. Disclosed, and coalescing is out of scope, so your call. The cheaper fix is not batch-specific though: the second save exists only because `wp_insert_comment()` writes `comment_meta` after the count, the reverse of `wp_update_comment()`. Bracketing the insert in `wp_defer_comment_counting()` collapses it to one.

**Comment by oaratovskyi** [comment]
@mordeth I've addressed your findings, please take another look

**Comment by mordeth** [comment]
Move handling is right on the normal paths. One regression, plus test cleanup.

**1 (the new move block counts the source twice and never counts the destination):** v3:625 and v1/v2:397 assume `wp_update_comment()` refreshed the persisted product. It does not. It captures the count target at comment.php:2679, before the public `wp_update_comment_data` filter at :2694, writes the row from the filtered data at :2724, then counts the pre-filter target at :2743. So a callback that redirects the review has Core count the prepared product and the new block count the original again. Hooks fire `[source, source]`, destination stays 0/0. Confirmed by running it. The stale destination is pre-existing; the second source recount, so a redundant save and a duplicate public hook, is new here.

**Proposal:** capture what Core will count before the update, then refresh whichever of `{ original, persisted }` it missed.

```php
$core_counted_product_id = isset( $prepared_args['comment_post_ID'] )
	? (int) $prepared_args['comment_post_ID']
	: $original_product_id;

foreach ( array_unique( array( $original_product_id, $current_product_id ) ) as $product_id_to_count ) {
	if ( $product_id_to_count !== $core_counted_product_id ) {
		wp_update_comment_count( $product_id_to_count );
	}
}
```

Normal and `product_id`-driven moves stay as they are. Worth a test with a redirecting callback asserting neither product is counted twice.

**Tests:**

- `test_update_item_rejects_malformed_filtered_comment_meta` and `..._rejects_scalar_filtered_review` differ only by the callback. `@dataProvider` halves them.
- The three move tests repeat the same source/destination/review setup. A helper next to `create_review()` carries it.
- The order assertion at :506 sits on top of aggregate assertions that already cover the outcome. Order-independent over the same IDs keeps the "exactly two recounts" part, and would have caught 1.
- The global `$post` rig at :538 is unreachable, `0 >= $product_id` short-circuits first. Fine as a guard, the comment just overclaims.

**Follow-ups (not blocking):**

- Create still saves twice, and no single-save version exists at this layer: `wp_insert_comment()` counts at :2148 and stores `comment_meta` at :2162, the inverse of the update path, so passing `comment_meta` there will not help. Disclosed, and matches the front-end path.
- The `comment_post_ID = 0` rejection is filter-level BC too, not only request validation: an extension detaching a review by returning `0` now gets a 404.
- Changelog says wc/v3 moves; v1 recounts the source now too.

**Comment by oaratovskyi** [comment]
@mordeth thanks for the careful review! I pushed the follow-up in https://github.com/woocommerce/woocommerce/pull/67778/commits/e36142668997e9cf073ab7f648fff3b27c6bb1bb and updated the changelog in https://github.com/woocommerce/woocommerce/pull/67778/commits/0ddc0b87fcd34b5013ea13bae5328df912cf52c0.

The recount path now captures the product Core intends to count before `wp_update_comment()` runs, re-reads the persisted review afterward, and recounts only the unique original/destination products Core missed. The new wc/v1 and wc/v3 `wp_update_comment_data` regressions verify that each product is counted exactly once and that both WooCommerce aggregates and core comment counts are current.

I also consolidated the malformed-filter cases with data providers, extracted the repeated wc/v3 move fixture, made recount assertions order-independent while still rejecting duplicates, and clarified the global-post zero-ID fixture comment. The rated-create save behavior remains unchanged.

I updated the PR description and changelog to reflect the complete wc/v1–v3 scope and explicitly called out the filter-level compatibility change: a wc/v3 preprocess filter returning `comment_post_ID = 0` now receives the same 404 as request input supplying zero.

Would you mind taking another look?

**Review by mordeth — APPROVED** [review]
Approving. The move-recount fix is correct: core captures its count target before `wp_update_comment_data` can redirect the review, and the new loop covers exactly the products core missed. Verified red-on-revert (both new tests fail with `[source, source]` when the production hunks are reverted), and the trunk merge is clean.

Four nits, none blocking:

1. `$core_counted_product_id` uses `??` where core uses `array_merge`. On a null `comment_post_ID` from a preprocess filter, core overwrites to null and its own count no-ops, while `??` falls back to `$original_product_id`, so the controller concludes core counted the source and the loop skips everything. Result is 200, rating stored, no aggregate refresh. Trunk and the previous head behaved the same way, so it is not a regression, but `array_key_exists()` closes it.

2. The changelog now reads "moved between products in wc/v1, wc/v2, and wc/v3". A client-initiated move is only reachable on v3: v1/v2 nest the route under `product_id` and `get_review()` enforces ownership, so those 404. v1/v2 moves are filter-driven only. The BC section already gets this right with "request- or filter-driven".

3. `test_update_item_status_changes_keep_the_aggregates_correct` is deleted with no replacement, so nothing exercises a status-only PUT now. I replayed it against this head and it passes, so this is coverage loss rather than a hidden defect. Worth restoring.

4. The description says each product is counted exactly once. That does not hold when a move and a status change arrive in the same PUT: core counts the destination inside `wp_update_comment()`, then `handle_status_param()` counts it again. Predates this PR, so it is just the sentence.

My earlier comments still stand for the empty-array filter drop, the BC bullet on falsy non-arrays, and the create-path `wp_defer_comment_counting()` bypass.

File: plugins/woocommerce/includes/rest-api/Controllers/Version1/class-wc-rest-product-reviews-v1-controller.php [hunk 4]
```diff
@@ -343,18 +356,54 @@ public function update_item( $request ) {
...
+
+		$core_counted_product_id = (int) ( $prepared_review['comment_post_ID'] ?? $original_product_id );
+
+		/*
+		 * Zero means no comment-table row changed; false is the actual update failure. A rating-only
+		 * request legitimately returns zero while still storing comment_meta and refreshing counts.
+		 */
 		$updated = wp_update_comment( $prepared_review );
-		if ( 0 === $updated ) {
+		if ( false === $updated ) {
```
Comment (tjcafferkey) [inline plugins/woocommerce/includes/rest-api/Controllers/Version1/class-wc-rest-product-reviews-v1-controller.php:388]: `$updated` can also be `WP_Error`

File: plugins/woocommerce/includes/rest-api/Controllers/Version3/class-wc-rest-product-reviews-controller.php [hunk 5]
```diff
@@ -586,8 +615,20 @@ public function update_item( $request ) {
 			}
 		}
 
-		if ( ! empty( $request['rating'] ) ) {
-			update_comment_meta( $id, 'rating', $request['rating'] );
+		// Re-read the comment because the same request can move the review to a different product.
+		$updated_review     = get_comment( $id );
```
Comment (tjcafferkey) [inline plugins/woocommerce/includes/rest-api/Controllers/Version3/class-wc-rest-product-reviews-controller.php:619]: We also `get_comment()` further down here too unconditionally but we only write to it conditionally. An optimisation we can make is only executing `get_comment()` again if we actually write to it, otherwise we can just return `$updated_review`

Comment (tjcafferkey) [reply plugins/woocommerce/includes/rest-api/Controllers/Version3/class-wc-rest-product-reviews-controller.php:619]: We should probably use `$this->get_review()` also for consistency?

Comment (tjcafferkey) [reply plugins/woocommerce/includes/rest-api/Controllers/Version3/class-wc-rest-product-reviews-controller.php:619]: Also I'm not sure the final `get_comment()` is actually needed? Might be worth double checking but I don't think the fields we are updating here get returned in `WP_Comment`

File: plugins/woocommerce/tests/php/includes/rest-api/Controllers/Version1/class-wc-rest-product-reviews-v1-controller-tests.php [hunk 6]
```diff
@@ -170,4 +170,497 @@ public function test_cannot_delete_other_comment_types() {
...
+		$product = wc_get_product( $product_id );
+		$this->assertEquals( 0, $product->get_average_rating() );
+		$this->assertSame( array(), $product->get_rating_counts() );
+		$this->assertEquals( 1, $product->get_review_count() );
+	}
+
+	/**
+	 * @testdox Editing the rating of a review recalculates the product rating aggregates.
+	 */
+	public function test_update_item_recalculates_the_aggregates_when_the_rating_changes() {
```
Comment (tjcafferkey) [inline plugins/woocommerce/tests/php/includes/rest-api/Controllers/Version1/class-wc-rest-product-reviews-v1-controller-tests.php]: Could probably cover this and `test_update_item_accepts_an_edit_that_changes_only_the_rating` in the same test?

File: plugins/woocommerce/tests/php/includes/rest-api/Controllers/Version3/class-wc-rest-product-reviews-controller-tests.php [hunk 7]
```diff
@@ -200,4 +200,475 @@ public function test_cannot_delete_other_comment_types() {
...
+		$product = wc_get_product( $product_id );
+		$this->assertEquals( 4, $product->get_average_rating(), 'An unrated review does not affect the average.' );
+		$this->assertEquals( array( 4 => 1 ), $product->get_rating_counts(), 'An unrated review is not counted as a rating.' );
+		$this->assertEquals( 2, $product->get_review_count(), 'An unrated review is still counted as a review.' );
+	}
+
+	/**
+	 * @testdox Editing the rating of a review recalculates the product rating aggregates.
+	 */
+	public function test_update_item_recalculates_the_aggregates_when_the_rating_changes() {
```
Comment (tjcafferkey) [inline plugins/woocommerce/tests/php/includes/rest-api/Controllers/Version3/class-wc-rest-product-reviews-controller-tests.php]: Looks like you can merge this test and the one below it maybe

**Review by tjcafferkey — CHANGES_REQUESTED** [review]
Thanks for this @oaratovskyi! I've left a few comments that are probably worth a look. Let me know what you think :)

File: plugins/woocommerce/includes/rest-api/Controllers/Version1/class-wc-rest-product-reviews-v1-controller.php (hunk 4 above)
Comment (oaratovskyi) [reply plugins/woocommerce/includes/rest-api/Controllers/Version1/class-wc-rest-product-reviews-v1-controller.php:388]: @tjcafferkey this one is not applicable. `wp_update_comment()` only returns `WP_Error` when its second argument is `true`; this call uses the default `false`, which normalizes failures to `false`.

File: plugins/woocommerce/includes/rest-api/Controllers/Version3/class-wc-rest-product-reviews-controller.php (hunk 5 above)
Comment (oaratovskyi) [reply plugins/woocommerce/includes/rest-api/Controllers/Version3/class-wc-rest-product-reviews-controller.php:619]: @tjcafferkey I checked `verified` is comment meta rather than a `WP_Comment` field, so WooCommerce itself doesn’t need the final re-read. However, that read currently happens after the public comment-meta hooks, and removing it entirely could prevent comment-row changes made by extensions from being reflected in the subsequent action and response. 

To keep this BC-safe, I’d go with your conditional option: reuse `$updated_review` when `verified` isn’t written, and retain `get_comment()` only after the `verified` update. I’d also keep `get_comment()` rather than switch to `get_review()`, since that could introduce a new `WP_Error` response after persistence. Does that sound reasonable?

Comment (oaratovskyi) [reply plugins/woocommerce/includes/rest-api/Controllers/Version3/class-wc-rest-product-reviews-controller.php:619]: Addressed in https://github.com/woocommerce/woocommerce/commit/89ad74feb0. WDYT?

File: plugins/woocommerce/tests/php/includes/rest-api/Controllers/Version3/class-wc-rest-product-reviews-controller-tests.php (hunk 7 above)
Comment (oaratovskyi) [reply plugins/woocommerce/tests/php/includes/rest-api/Controllers/Version3/class-wc-rest-product-reviews-controller-tests.php]: Good call here too 🙂  Done in https://github.com/woocommerce/woocommerce/commit/89ad74feb0.

File: plugins/woocommerce/tests/php/includes/rest-api/Controllers/Version1/class-wc-rest-product-reviews-v1-controller-tests.php (hunk 6 above)
Comment (oaratovskyi) [reply plugins/woocommerce/tests/php/includes/rest-api/Controllers/Version1/class-wc-rest-product-reviews-v1-controller-tests.php]: Good call! Consolidated this into the rating-only regression test in https://github.com/woocommerce/woocommerce/commit/89ad74feb0.

**Comment by oaratovskyi** [comment]
@tjcafferkey, thanks for the review! I have handled all four items. Could you take another look and let me know if any items are still blocking?

**Review by tjcafferkey — APPROVED** [review]
Thanks for addressing the feedback @oaratovskyi - this now LGTM!

**Comment by oaratovskyi** [comment]
No docs update needed here — noting it for anyone scanning the thread later.

This PR only changes the bodies of `create_item()` and `update_item()` in the two review controllers. It doesn't touch `register_routes()`, `get_item_schema()`, or `get_collection_params()`, so everything the REST API docs describe is still accurate: the product review properties table, the endpoint paths, the example payloads, and the schema-derived route index in `v3/api-reference.mdx`.

The behavior that did change — rating aggregates recomputing in the right order, and an invalid wc/v3 destination returning `woocommerce_rest_product_invalid_id` — isn't something those pages cover. They document no error codes today, so adding the new 404 would mean establishing an error-response convention for the page rather than correcting a stale doc.
